### PR TITLE
Security Log

### PR DESCRIPTION
**Pull request by simeonbizhub**
[Trello](https://trello.com/c/EVsMrsQb)

These changes introduce a new `SecurityLog` document in the admin module, similar to `Audit`.
A `SecurityLog` instance can be created in SecurityUtil, by either specifying an exception, or a custom 'event type' and 'event message.'
A `SecurityLog` instance is created automatically in the constructor of `SecurityException`, when `AccessException` or `SecurityException` are thrown. 
In order to accomodate these changes, there was some re-jigging required to manage dependencies.

**Re-jigging**
- Removed `HttpServletRequest` & `HttpServletResponse` from `WebContext`
- Added `HttpServletRequestResponse` 
- Added `WebContainer` class to set/get `HttpServletRequestResponse`
- Moved `User.checkAccess` to `EXT` method
- Moved `AccessException`, `SecurityException` from skyve-core to skyve-ext
- Made `HttpServletRequestResponse` injectable via `SkyveCDIProducer`
- Added `getHttpServletRequestResponse` to `EXT`
- Update to `RequestLoggingAndStatisticsFilter` to set request/response (cleared in try-finally)

**Changes**
- Added `SecurityLog` document to admin module
- Added `SecurityLog` interface in skyve-core (`admin.SecurityLog` implements)
- Added `SecurityUtil`
- Implemented `SecurityUtil.log` for both exception & custom declarations
- Implemented `SecurityUtil.email`
- Added `RecentSecurityLogModel` to `SystemDashboard`

### CODE REVIEW

File: skyve-core/src/main/java/org/skyve/domain/app/admin/SecurityLog.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.skyve.domain.PersistentBean;
+import org.skyve.domain.app.AppConstants;
+import org.skyve.domain.messages.DomainException;
+import org.skyve.domain.types.Timestamp;
+
+public interface SecurityLog extends PersistentBean {
+
+	public static SecurityLog newInstance() {
+		try {
+			return CORE.getUser()
```
Comment (sandsm) [inline skyve-core/src/main/java/org/skyve/domain/app/admin/SecurityLog.java]: 3 Calls to CORE.getUser() here, 2 to getCustomer()

File: skyve-core/src/main/java/org/skyve/domain/app/admin/SecurityLog.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+					.getCustomer()
+					.getModule(AppConstants.ADMIN_MODULE_NAME)
+					.getDocument(CORE.getUser().getCustomer(), AppConstants.SECURITY_LOG_DOCUMENT_NAME)
+					.newInstance(CORE.getUser());
+		} catch (RuntimeException e) {
```
Comment (sandsm) [inline skyve-core/src/main/java/org/skyve/domain/app/admin/SecurityLog.java]: Do we need to catch Exceptions here? Only throw if you need to, probs chain with a description no matter if its runtime or not - ie ("Could not instantiate a new SecurityLog", e)

File: skyve-ext/src/main/java/org/skyve/EXT.java [hunk 3]
```diff
@@ -655,5 +661,89 @@ public static void bootstrap(@Nonnull Persistence p) throws Exception {
...
+	}
+
+	/**
+	 * Get the {@link HttpServletRequest} & {@link HttpServletResponse} from the current transaction.
+	 * <br/>
+	 * Can return null if called from a job or other background task.
+	 * 
+	 * @return An object containing both the request & response
+	 */
+	public static HttpServletRequestResponse getHttpServletRequestResponse() {
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/EXT.java]: Add @Nullable on return type

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+public class SecurityUtil {
+
+	/**
+	 * Creates a {@link SecurityLog} entry and emails its contents to the defined support user.
+	 * 
+	 * @param exception The exception raised for this security event
+	 */
+	public static void log(@Nonnull Exception exception) {
+		String eventType = exception.getClass()
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:28]: Maybe a comment for what this is doing?

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	/**
+	 * Creates a {@link SecurityLog} entry and emails its contents to the defined support user.
+	 *
+	 * @param eventType The type of security event
+	 * @param eventMessage What is this security event
+	 * @param exception Optionally provide an exception to extract provenance
+	 * 
+	 * @author Simeon Solomou
+	 */
+	private static void log(@Nonnull String eventType, @Nonnull String eventMessage, Exception exception) {
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java]: Is Exception @Nonnull?

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+				// Event message
+				sl.setEventMessage(eventMessage);
+
+				if (exception != null) {
+					// Provenance
+					sl.setProvenance(getProvenance(exception));
+				}
+
+				try {
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:112]: Do you think its appropriate to chain these exceptions?

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	 * @param sl The newly created security log
+	 * @author Simeon Solomou
+	 */
+	private static void email(SecurityLog sl) {
+		String supportEmail = UtilImpl.SUPPORT_EMAIL_ADDRESS;
+		if (supportEmail == null) {
+			Util.LOGGER.warning("Cannot send security log notification as no support email address is specified");
+			return;
+		}
+		if ("localhost".equals(UtilImpl.SMTP)) {
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:162]: Are you sure we should be checking for localhost?
It is feasible there is a setup on a VPS which has a linux mail server running.

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+				.body(body.toString()));
+	}
+
+	/**
+	 * Returns the source IP for the parsed {@link HttpServletRequest}
+	 * 
+	 * @param request
+	 * @return source IP
+	 */
+	public static String getSourceIpAddress(HttpServletRequest request) {
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:193]: There is also a standard "Forwarded" header to inspect - Maybe this method should go in WebUtil.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded
Prefer the standard one before looking for the X-Forwarded-... stuff

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+		String xForwardedForHeader = request.getHeader("X-Forwarded-For");
+		if (xForwardedForHeader == null) {
+			return request.getRemoteAddr();
+		}
+
+		// As of https://en.wikipedia.org/wiki/X-Forwarded-For
+		// The general format of the field is: X-Forwarded-For: client, proxy1, proxy2 ...
+		// we only want the client
+		return new StringTokenizer(xForwardedForHeader, ",").nextToken()
```
Comment (sandsm) [inline skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java]: nextToken could throw - guard that or use asIterator().

File: skyve-war/src/main/java/modules/admin/ReportDesign/actions/Preview.java [hunk 10]
```diff
@@ -56,7 +57,9 @@ public Download download(ReportDesign bean, WebContext webContext) throws Except
 
 		final Map<String, Object> parameters = new HashMap<>();
 
-		final UxUi uxui = UserAgent.getUxUi((HttpServletRequest) webContext.getHttpServletRequest());
+		HttpServletRequest request = EXT.getHttpServletRequestResponse()
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/ReportDesign/actions/Preview.java]: Potential null problems?

File: skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml [hunk 11]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<document xmlns="http://www.skyve.org/xml/document"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	name="SecurityLog"
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml]: Name element first?

File: skyve-war/src/generated/java/modules/admin/domain/SecurityLog.java [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+	 * When the exception was raised
+	 **/
+	private Timestamp timestamp = (Timestamp) ExpressionEvaluator.evaluate("{TIMESTAMP}", this);
+
+	/**
+	 * Thread ID
+	 * <br/>
+	 * Thread ID when the exception was raised
+	 **/
+	private Long threadID;
```
Comment (sandsm) [inline skyve-war/src/generated/java/modules/admin/domain/SecurityLog.java]: Investigate using the Srting thread identifier used in wildfly to easily marry up the log with the security exception?

File: skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+		<id name="loggedInUserId" required="true">
+			<displayName>Logged In User ID</displayName>
+			<description>Logged in user ID when exception was raised</description>
+		</id>
+		<text name="eventType" required="true">
+			<displayName>Event Type</displayName>
+			<description>Type of security event raised. If raised with an exception, this is the exception name.</description>
+			<length>500</length>
+		</text>
+		<memo name="eventMessage" required="true">
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml:59]: This is text indexed by default - should it be?

File: skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+			<displayName>Event Message</displayName>
+			<description>Details about the security event. If an exception, this is the message from raised exception</description>
+		</memo>
+		<memo name="provenance">
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/SecurityLog.xml:64]: This is text indexed by default - should it be

File: skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml [hunk 15]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<view xmlns="http://www.skyve.org/xml/view"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="edit"
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml]: name and title elements first.

File: skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	title="{bizKey}"
+	xsi:schemaLocation="http://www.skyve.org/xml/view ../../../../schemas/view.xsd">
+	<form border="true" borderTitle="Security Log Details" disabled="true">
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml:7]: indentation looks wrong here

File: skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+            	<blurb>{eventMessage}</blurb>
+            </item>
+        </row>
+         <row>
+            <item showLabel="true" label="Provenance">
+            	<blurb>{provenance}</blurb>
+            </item>
+        </row>
+    </form>
+	<actions>
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SecurityLog/views/edit.xml:57]: indentation looks wrong here

File: skyve-war/src/main/java/modules/admin/SelfRegistration/actions/Register.java [hunk 18]
```diff
@@ -33,14 +33,17 @@ public class Register implements ServerSideAction<SelfRegistrationExtension> {
 	@Override
 	public ServerSideActionResult<SelfRegistrationExtension> execute(SelfRegistrationExtension bean, WebContext webContext) throws Exception {
 		Persistence persistence = CORE.getPersistence();
+		HttpServletRequest request = EXT.getHttpServletRequestResponse()
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SelfRegistration/actions/Register.java]: Potential Null problem possible?

File: skyve-war/src/main/java/modules/admin/SelfRegistrationActivation/SelfRegistrationActivationBizlet.java [hunk 19]
```diff
@@ -12,7 +13,8 @@ public class SelfRegistrationActivationBizlet extends Bizlet<SelfRegistrationAct
 	public SelfRegistrationActivationExtension preExecute(ImplicitActionName actionName, SelfRegistrationActivationExtension bean, Bean parentBean, WebContext webContext)
 			throws Exception {
 		if (ImplicitActionName.New.equals(actionName)) {
-			HttpServletRequest request = (HttpServletRequest) webContext.getHttpServletRequest();
+			HttpServletRequest request = EXT.getHttpServletRequestResponse()
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SelfRegistrationActivation/SelfRegistrationActivationBizlet.java]: NPE?

File: skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
+package modules.admin.SystemDashboard.models;
```
Comment (sandsm) [inline skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java:1]: Is the only reason this model exists to show 5?

**Review by sandsm — CHANGES_REQUESTED** [review]
Nice work @simeonbizhub.
Comments to address enclosed.
Some of them are genuine questions coz I dont have the benefit of the compiler during review, so they are just hunch - ignore if I'm wrong.

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java (hunk 5 above)
Comment (simeonbizhub) [reply skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java]: I have changed this parameter to `@Nullable String provenance`, as this is all that parameter is used for here.

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java (hunk 6 above)
Comment (simeonbizhub) [reply skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:112]: My approach to this section of code was to try assure that each part of the process (upsert, email, commit, close etc.) run independently of each other. For example, if the SecurityLog save fails, an email can still be sent with its contents.

I am not throwing any of these exceptions, just logging.  I don't think any of these exceptions can chain, as each piece of code _can_ be run independently - am I missing something or misinterpreting?

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java (hunk 7 above)
Comment (simeonbizhub) [reply skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:162]: Good point - I just copied the code from `ConfigurationExtension.validSMTPHost` here.

Is there a better way for me to check that a valid email has been configured before sending?

File: skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java (hunk 8 above)
Comment (simeonbizhub) [reply skyve-ext/src/main/java/org/skyve/util/SecurityUtil.java:193]: Updated to check "Forwarded" header first.

Would be good in WebUtil, but SecurityUtil would not have access to it there as we are in skyve-ext.

File: skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java (hunk 20 above)
Comment (simeonbizhub) [reply skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java:1]: Yes - is there a simpler way to achieve this without creating a model? 
Or are there more customisations you would like here?

File: skyve-war/src/generated/java/modules/admin/domain/SecurityLog.java (hunk 12 above)
Comment (simeonbizhub) [reply skyve-war/src/generated/java/modules/admin/domain/SecurityLog.java]: I added 'Thread Name' attribute to SecurityLog - `Thread.currentThread.getName`.
Is this what you mean?

File: skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java (hunk 20 above)
Comment (sandsm) [reply skyve-war/src/main/java/modules/admin/SystemDashboard/models/RecentSecurityLogModel.java:1]: I guess maybe a collection and data grid for 5 but the benefits of the model is that it is lazy loaded outside the request response and only loaded on tab change (in SC) so its a better technical solution.

**Review by sandsm — APPROVED** [review]

**Pull request merged**